src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.OSX.cs

```diff
@@ -0,0 +1,233 @@
+                case HashAlgorithmNames.SHA384:
+                    return new AppleDigestProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha384);
+                case HashAlgorithmNames.SHA512:
+                    return new AppleDigestProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha512);
+            }
+
+            throw new PlatformNotSupportedException();
+        }
+
+        public static HashProvider CreateMacProvider(string hashAlgorithmId, byte[] key)
+        {
+            switch (hashAlgorithmId)
+            {
+                case HashAlgorithmNames.MD5:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Md5, key);
+                case HashAlgorithmNames.SHA1:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha1, key);
+                case HashAlgorithmNames.SHA256:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha256, key);
+                case HashAlgorithmNames.SHA384:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha384, key);
+                case HashAlgorithmNames.SHA512:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha512, key);
+            }
+
+            throw new PlatformNotSupportedException();
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private class AppleHmacProvider : HashProvider
+        {
+            private readonly byte[] _key;
+            private readonly SafeHmacHandle _ctx;
+
+            private bool _running;
+
+            public override int HashSizeInBytes { get; }
+
+            internal AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm algorithm, byte[] key)
+            {
+                _key = key.CloneByteArray();
+                int hashSizeInBytes = 0;
+                _ctx = Interop.AppleCrypto.HmacCreate(algorithm, ref hashSizeInBytes);
+
+                if (hashSizeInBytes < 0)
+                {
+                    _ctx.Dispose();
+                    throw new PlatformNotSupportedException(
+                        SR.Format(
+                            SR.Cryptography_UnknownHashAlgorithm,
+                            Enum.GetName(typeof(Interop.AppleCrypto.PAL_HashAlgorithm), algorithm)));
+                }
+
+                if (_ctx.IsInvalid)
+                {
+                    throw new CryptographicException();
+                }
+
+                HashSizeInBytes = hashSizeInBytes;
+            }
+
+            public override unsafe void AppendHashDataCore(byte[] data, int offset, int count)
+            {
+                if (!_running)
+                {
+                    SetKey();
+                }
+
+                int ret;
+
+                fixed (byte* pData = data)
```

Comment:
I was just having that debate internally.  It looks like all 3 implementations have the same approach; so I'm going to poke at this a bit.  Maybe the parent function should be guarded with `if (count == 0) { return; }`